File: frontend/src/types/productDispositionStock.ts
/** Read-only disposition pools from GET /products (Etap 1 — additive API). */
export type ProductDispositionStock = {
  saleable_qty: number;
  outlet_qty: number;
  service_qty: number;
  quarantine_qty: number;
  scrap_qty: number;
  rejected_qty: number;
  other_qty: number;
  physical_qty: number;
  saleable_available_qty: number;
};

export const EMPTY_DISPOSITION_STOCK: ProductDispositionStock = {
  saleable_qty: 0,
  outlet_qty: 0,
  service_qty: 0,
  quarantine_qty: 0,
  scrap_qty: 0,
  rejected_qty: 0,
  other_qty: 0,
  physical_qty: 0,
  saleable_available_qty: 0,
};

/** Future Etap 2: OrderItem.required_stock_disposition will use these codes. */
export const CANONICAL_STOCK_DISPOSITIONS = [
  "SALEABLE",
  "OUTLET_B",
  "SERVICE_C",
  "QUARANTINE",
  "SCRAP",
  "REJECTED_STOCK",
] as const;

export type CanonicalStockDisposition = (typeof CANONICAL_STOCK_DISPOSITIONS)[number];

export function parseDispositionStock(raw: unknown): ProductDispositionStock | undefined {
  if (!raw || typeof raw !== "object") return undefined;
  const o = raw as Record<string, unknown>;
  const num = (k: string) => {
    const v = o[k];
    return v != null && Number.isFinite(Number(v)) ? Number(v) : 0;
  };
  return {
    saleable_qty: num("saleable_qty"),
    outlet_qty: num("outlet_qty"),
    service_qty: num("service_qty"),
    quarantine_qty: num("quarantine_qty"),
    scrap_qty: num("scrap_qty"),
    rejected_qty: num("rejected_qty"),
    other_qty: num("other_qty"),
    physical_qty: num("physical_qty"),
    saleable_available_qty: num("saleable_available_qty"),
  };
}

export function fmtDispositionQty(n: number): string {
  const r = Math.round(n);
  if (Math.abs(n - r) < 1e-9) return String(r);
  return new Intl.NumberFormat("pl-PL", { maximumFractionDigits: 2 }).format(n);
}

File: backend/schemas/product_disposition_stock.py
"""Read-only product stock breakdown by warehouse disposition (Etap 1 — additive API)."""

from __future__ import annotations

from pydantic import BaseModel, Field


class ProductDispositionStockOut(BaseModel):
    """
    Physical on-hand per disposition pool (SSOT for product card / list).

    - ``saleable_available_qty`` = pick-eligible A − A reservations (default ATP / marketplace)
    - ``outlet_available_qty`` = pick-eligible B − B reservations (explicit outlet only)
    - ``service_available_qty`` = pick-eligible C (never reservable)
    - ``physical_qty`` = A+B+C+… (legacy stock_quantity parity)
    """

    saleable_qty: float = Field(0, ge=0, description="On-hand SALEABLE (A)")
    outlet_qty: float = Field(0, ge=0, description="On-hand OUTLET_B (B)")
    service_qty: float = Field(0, ge=0, description="On-hand SERVICE_C (C)")
    quarantine_qty: float = Field(0, ge=0, description="On-hand QUARANTINE")
    scrap_qty: float = Field(0, ge=0, description="On-hand SCRAP — not sellable/reservable")
    rejected_qty: float = Field(0, ge=0, description="On-hand REJECTED_STOCK — separate from quarantine")
    other_qty: float = Field(0, ge=0, description="Unknown disposition codes (normalized fallback bucket)")
    physical_qty: float = Field(0, ge=0, description="Sum of all visible inventory rows (= legacy stock_quantity)")
    saleable_available_qty: float = Field(
        0,
        ge=0,
        description="max(0, pick-eligible saleable_qty - SALEABLE reserved); excludes DOCK when requires_putaway",
    )
    outlet_available_qty: float = Field(
        0,
        ge=0,
        description="max(0, pick-eligible outlet_qty - OUTLET_B reserved)",
    )
    service_available_qty: float = Field(
        0,
        ge=0,
        description="Pick-eligible SERVICE_C on-hand (not reservable)",
    )
    dock_qty: float = Field(
        0,
        ge=0,
        description="On-hand SALEABLE on DOCK-IN (physical buffer awaiting putaway)",
    )
